pkg/pgmodel/Readme.md

```diff
@@ -36,46 +80,51 @@ type Sample {
-3. Before forwarding the batch to the next stage, replaces all `Labels` with `SeriesIds` fetches missing `SeriesIds` from the DB.
+The ingestor accepts a `WriteRequest` at `ingestor.IngestMetrics(..., *WriteRequest)`. The ingestor determines if it needs to ingest `TimeSeries` only, or also `MetricMetadata` (which we were ignoring).
 
-The purpose of this metric layer is to allow for greater batching series-id fetches, and metric table `INSERT`s both of which perform notably better if they work on larger amounts of data.
+To process a `TimeSeries`, the ingestor:   
+1. Transforms the Labels field (a `{string, string}[]`) into a `model.Series`, (which it gets from the series cache).
+2. Groups the `model.Series` together with its `[]Samples` as a `model.Insertable`.
+3. Puts all `model.Insertable` entries into a map, keyed by the metric name (`map[string][]model.Insertable`)
 
-After a batch is finished by the metric layer its forwarded to an arbitrary physical inserter, which translates the insert request into the actual SQL doing the insert, sends the request to the DB, and reports the result back to the original goroutine. We maintain a fixed number of physical inserters based on the number of connections to the DB we possess.
+The ingestor passes this map on to the dispatcher (`dispatcher.InsertTs(...)`).
 
-Diagrammatically the write path cn be visualized as follows:
+#### Dispatcher
 
-```
-|, /, \ are go channels
-= is for network connections
-- is for normal function calls
+The dispatcher processes this map by iterating through the `(key, value)` pairs (which are metric name, and the insertable (`model.Insertable`)), and sends it on a channel to the one metric batcher goroutine responsible for batching samples for that metric. If the metric batcher does not exist yet, it is created on demand.
 
-                  new req
-                     v
-                 +--------+        +-------------+
-                 | Ingest | <----> | Label Cache |
-                 +--------+        +-------------+
-                  /     \
-                 /       \
- +----------------+     +----------------+       +-----------------+
- | metric handler | ... | metric handler | <===> | DB (series IDs) |
- +----------------+     +----------------+       +-----------------+
-                 \       /
-                  \     /
-                   \   /
-                    \ /
- +----------------+  |  +----------------+       +------------------+
- | insert handler | ... | insert handler | ====> | DB (metric data) |
- +----------------+     +----------------+       +------------------+
-```
+If asyncAcks=true, the dispatcher returns immediately, otherwise it waits for the metric batcher goroutine to return a result, and returns that in response.
+
+### Metric batcher
+
+The second stage consists of the metric batcher, one goroutine for each metric name. These goroutines are created in `dispatcher.InsertTs`, but are pooled for later re-use.
```

Comment:
I think this might be a bit misleading as someone might think there is a pool of batchers which are re-used. Essentially we just have a batcher per metric name running in it's own goroutine and communicating with dispatcher and copiers using channels.

Reply:
done